Extract some player attributes to data files

glebm opened the pull request.

glebm commented:
@ephphatha The header parsing allows the columns to be in any order but do we actually need it?
I think a `skipHeader()` function would be fine for most of these files

ephphatha commented:
> @ephphatha The header parsing function allows the columns to be in any order but do we actually need it? I think a `skipHeader()` function would be fine for most of these files. I can add one in a follow-up PR

You can call `parseHeader()` with an empty range (even nullptr, nullptr), so skipHeader could be a simple wrapper for that.

The column heading order isn't as important in a transposed table like this PR. As a followup to https://github.com/diasurgical/devilutionX/pull/6530 I was going to effectively lift and shift the arrays to tab-separated files, which would make it more useful (one of the things commonly seen in d2 text files are comment columns which we would want to skip anyway, allowing reordering wasn't particular difficult to achieve once that was implemented).

I would be interested in seeing a fixed point equivalent to `std::from_chars()` for float so we don't have to get people to specify raw values, I haven't had time to dig into that myself but that was really the only thing I was considering a blocker before loading life/mana bonuses from files.

ephphatha commented on Source/playerdat.cpp:
```diff
@@ -152,11 +155,122 @@ void ReloadExperienceData()
+	for (DataFileRecord record : dataFile) {
+		FieldIterator fieldIt = record.begin();
+		const FieldIterator endField = record.end();
+		if (fieldIt == endField) {
```
This comparison will never be true, `fieldIt` will always contain at least one (potentially zero length) field so will never compare equal to the sentinel until an increment is attempted. Even if you perform a value access this isn't enough to cause the iterator to reach the sentinel value, it has to be incremented.

glebm replied:
I see, makes sense, removed

ephphatha commented on Source/playerdat.cpp:
```diff
@@ -152,11 +155,122 @@ void ReloadExperienceData()
+		        return tl::make_unexpected(ColumnDefinition::Error::UnknownColumn);
+	        });
+	    !result.has_value()) {
+		DataFile::reportFatalError(result.error(), filename);
```
With the way you're currently handling values you'll want to report an error here if both columns are present but not in the expected order.

glebm replied:
There is no expected order, it's an `unordered_map`

glebm replied:
Ah, you mean columns, yeah.
I've changed it to simply `skipHeader()`

glebm commented:
Thanks, I've added a simple `skipHeader` function based on the `parseHeader` one but much simpler.

As for having floats in data files, I think it's fine to just have ints. Floats introduce a whole other layer of complexity.

glebm commented:
Perhaps we can simplify it further by mandating the field order as well (with validation). There isn't any need to reorder them and mods can add values to the end. What do you think?

glebm commented:
Pushed the code mandating the field order in a separate commit

ephphatha commented:
> Perhaps we can simplify it further by mandating the field order as well (with validation). There isn't any need to reorder them and mods can add values to the end. What do you think?

Part of the aim of having arbitrary column ordering was to allow mods to group related columns to make files easier to read/modify. I can see the reason for transposing the tables and having a per-class split however I must admit I'm even less keen about having a strict row order. That said it doesn't matter what my preference is, it matters more what modders find easy to use :D

ephphatha commented on Source/playerdat.cpp:
```diff
@@ -152,11 +153,94 @@ void ReloadExperienceData()
+
+		++fieldIt;
+		if (fieldIt == endField) {
+			DataFile::reportFatalError(DataFile::Error::NotEnoughColumns, filename);
```
Oh, turns out in the WIP branch I hadn't pushed yet I do exactly this, create a DataFileField::Error enum and use it instead of std::errc in reportFatalFieldError

ephphatha replied:
Also just realised I never cleaned up the use of this when parsing experience data... no need to worry about it in this one since there's already convention.

glebm commented on Source/playerdat.cpp:
```diff
@@ -152,11 +153,94 @@ void ReloadExperienceData()
+
+		const std::string_view key = (*fieldIt).value();
+		if (key != expectedKey) {
+			app_fatal(fmt::format("Unexpected field in {}: got {}, expected {}", filename, key, expectedKey));
```
I agree but let's only make them translatable once the error types have stabilized a bit, to reduce translation churn for now.

glebm commented:
The mods would have to add code that reads the new values, and it's much simpler to write code that expects the values in the particular order. In the order-independent code I had a map and a variant, it was rather complicated compared to this version. So this might actually be easier for modders, though I'm also not sure.

ephphatha commented:
Fair enough, I'm happy to go with this approach. It'll conflict with the initial scope of #6530 since you've used a slightly different API for exposing the PlayerData struct but I can rebase that work easy enough. I finished off the class-per-row parsing as an example of how that would be used but it's really not that much different to this approach, just means that the file structure is slightly different and the boilerplate to handle columns shifts to strict row order.

Only real hesitation I have is that I would really prefer allowing decimal representations of fixed point numbers instead of the raw 1/64 values. The fixed point parsing turned out to be fairly simple to implement so I've added that to my PR, I can cherry pick it for use here if you'd like.

glebm commented:
Yeah, the fixed point parsing turned out simpler than I thought, sounds good

glebm commented:
What would be the best way to merge the 2 PRs?

As I understand:
1. Fixed point :+1: 
2. Fixed order of field :+1: 
3. One file per class vs one file -- undecided?

glebm commented:
One advantage of having per-class files is that it makes it easier to add a new class -- simply copy the directory of another class and modify the values. If all classes are in the same files, one needs to then figure out which files relate to classes and manually add a row in each

ephphatha commented:
> What would be the best way to merge the 2 PRs?
> 
> As I understand:
> 
> 1. Fixed point 👍
> 2. Fixed order of fields / columns 👍
> 3. Files per class vs all classes in the same file -- undecided?

Files per class sounds fine to me. Regarding merging the PRs the only thing we're "missing" from this PR is the extracted toHit values and starting items, it's easy enough to roll back the scope of the one I had opened and rebase it on what you've done here.

glebm commented:
Done!

ephphatha commented on Packaging/resources/assets/txtdata/classes/README.md:
```diff
@@ -0,0 +1,25 @@
+ `maxDex`     | Class Maximum Dexterity Stat
+ `maxVit`     | Class Maximum Vitality Stat
+ `blockBonus` | Class Block Bonus, %
+ `adjLife`    | Class Life Adjustment, 1/64
+ `adjMana`    | Class Mana Adjustment, 1/64
+ `lvlLife`    | Life gained on level up, 1/64
+ `lvlMana`    | Mana gained on level up, 1/64
+ `chrLife`    | Life from base Vitality, 1/64
+ `chrMana`    | Mana from base Magic, 1/64
+ `itmLife`    | Life from item bonus Vitality, 1/64
+ `itmMana`    | Mana from item bonus Magic, 1/64
```
s|1/64|decimal value| (or something along those lines  :D)

glebm replied:
Done

ephphatha commented on Source/playerdat.cpp:
```diff
@@ -152,11 +153,99 @@ void ReloadExperienceData()
+	const auto readRecord = [&](std::string_view expectedKey, auto &outValue) {
+		DataFileField valueField = getValueField(expectedKey);
+		tl::expected<void, devilution::DataFileField::Error> parseIntResult;
+		if constexpr (std::is_same_v<std::remove_cvref_t<decltype(outValue)>, int16_t>) {
```
Is it possible to use this with a typedef for a fixed_16_6 type?  to prevent any surprises if someone changes one of the stat fields to int16_t for example

glebm replied:
Done

AJenbo commented:
Nice :)
